engine/engine/content/builtins/scripts/edn.lua

```diff
@@ -118,53 +121,90 @@ local type_to_primitive_encoder = {
   ["function"] = encode_function,
   ["thread"] = encode_thread
 }
+local metatable_to_primitive_encoder = {
+  [GOScriptInstance] = encode_script,
+  -- [GuiScriptInstance] = encode_script, (it doesn't work for *.gui_script)
+  [RenderScriptInstance] = encode_script,
+  [NodeProxy] = encode_node
+}
 
-local function fail_on_unexpected_type(x)
-  error("unexpected type '" .. type(x) .. "'")
+local function fail_on_unexpected_type(x, val_type)
+  error("unexpected type '" .. val_type .. "'")
 end
 
 local function encode_as_primitive(val)
-  local encoder = type_to_primitive_encoder[type(val)] or fail_on_unexpected_type
-  return encoder(val)
+  local val_type = type(val)
+  local mt = debug.getmetatable(val)
+  local encoder = (mt and metatable_to_primitive_encoder[mt.__metatable]) or type_to_primitive_encoder[val_type] or fail_on_unexpected_type
+  return encoder(val, val_type, mt)
 end
 
-local function encode_table(val)
+local function encode_edn_table(val, key)
   local encoded_kvs = {}
   for k, v in pairs(val) do
-    encoded_kvs[#encoded_kvs + 1] = encode_as_primitive(k) .. " " .. encode_as_primitive(v)
+    encoded_kvs[#encoded_kvs + 1] = encode_as_primitive(k)
+    encoded_kvs[#encoded_kvs + 1] = encode_as_primitive(v)
   end
-  return "#lua/table{:content [" .. table.concat(encoded_kvs, " ") .. "] :string " .. encode_string(str(val)) .. "}"
+  local encoded_str = encode_string(str(key, type(key), debug.getmetatable(key)))
+  return "#lua/table{:content [" .. table.concat(encoded_kvs, " ") .. "] :string " .. encoded_str .. "}"
 end
 
-local function encode_refs(refs)
+local function encode_refs(refs, keys)
   local encoded_refs = {}
   local val
   for i = 1, #refs do
     val = refs[i]
-    encoded_refs[i] = encode_as_primitive(val).." "..encode_table(val)
```

Comment:
We call `table.concat()` later where we use " " as separator. We don't need to have these concatinations too